### PR TITLE
[NPM] Generate TLS certificates during docker build and bake into image for gRPC secure channels

### PR DESCRIPTION
**Pull request by nitishm**
This change adds the generation of the Server cert and CA cert to the docker image. These certs are available to both the client and the server side to establish a secure gRPC connection between client & server.

Changes include,
- Adding tls helper functions to transport package
- Generate script to generate the TLS certificates
- Dockerfile (Linux) updated to run the generate script at build time.

TODOs,
- Create a generate script for Windows.

**Reason for Change**:

**Issue Fixed**:

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests

**Notes**:

### CODE REVIEW

File: npm/pkg/transport/tls.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	"os"
+
+	"google.golang.org/grpc/credentials"
+)
+
+const (
+	serverCertPEMFilename = "tls.crt"
+	serverKeyPEMFilename  = "tls.key"
+	caCertPEMFilename     = "ca.crt"
+	path                  = "/usr/local/npm"
```
[the remark this answers is not in the document]
Comment (nitishm) [reply npm/pkg/transport/tls.go:16]: The certs are baked in through the Dockerfile. Not sure how to make that dynamic. Hence I made this static :/

File: .gitignore [hunk 2]
```diff
@@ -24,3 +24,10 @@ ipam-*.xml
 controller-gen
 build/tools/bin
 npm/debug/http
+
+# certificates
+*/**/certs/
+*.crt
+*.pem
+*.pem
```
Comment (huntergregory) [inline .gitignore:32]: nit: duplicate

File: npm/pkg/transport/events_client.go [hunk 3]
```diff
@@ -81,11 +90,12 @@ func (c *EventsClient) run(ctx context.Context, stopCh <-chan struct{}) error {
 		default:
 			if connectClient == nil {
 				klog.Info("Reconnecting to gRPC server controller")
-				opts := []grpc.CallOption{grpc.WaitForReady(true)}
+				opts := []grpc.CallOption{grpc.WaitForReady(false)}
```
Comment (huntergregory) [inline npm/pkg/transport/events_client.go:93]: curious what this represents and why we had it true before, false now

File: npm/scripts/generate_certs.sh [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+if [ ! -f "$SAN_CNF_FILE" ]
+then
+		echo "SAN_CNF_FILE does not exist"
+		exit
+fi
+
+if [ ! -d "$CERTS_STAGING_DIR" ]
+then
+    echo "Creating $CERTS_STAGING_DIR"
+		mkdir -p $CERTS_STAGING_DIR
```
Comment (huntergregory) [inline npm/scripts/generate_certs.sh:25]: nit: inconsistent tab spacing (in above if statement too)

File: npm/Dockerfile [hunk 5]
```diff
@@ -20,6 +20,9 @@ FROM docker.io/ubuntu:focal
 COPY --from=builder /usr/local/bin/azure-npm \
 	/usr/bin/azure-npm
 
+COPY --from=builder /usr/local/src/npm/npm/scripts \
```
Comment (huntergregory) [inline npm/Dockerfile:23]: is `npm/npm/` intentional?

File: npm/cmd/root.go [hunk 6]
```diff
@@ -12,6 +19,27 @@ func NewRootCmd() *cobra.Command {
 		CompletionOptions: cobra.CompletionOptions{
 			DisableDefaultCmd: true,
 		},
+		PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
```
Comment (huntergregory) [inline npm/cmd/root.go:22]: why move this from the start cmd? Also has some impacts like to npm/Dockerfile, so might be better to scope to another PR?

Comment (nitishm) [reply npm/cmd/root.go:22]: What Dockerfile changes do you anticipate @huntergregory ?

File: npm/Dockerfile (hunk 5 above)
Comment (nitishm) [reply npm/Dockerfile:23]: Yep. Since I didnt want to change the WORKDIR in the builder image ([here](https://github.com/Azure/azure-container-networking/pull/1262/files/143d39929ebbd7773bb3a2a1662bc1c8137ec331#diff-9109d573bc356b2f5cfb6ccfa3a9a94643911cb8ef8e6b9de1e3a40debcf26f3R8-R11)) which copies the host's source to `/usr/local/src/npm` it has to be this way.

File: npm/pkg/transport/events_client.go (hunk 3 above)
Comment (nitishm) [reply npm/pkg/transport/events_client.go:93]: True blocks the connection until the server is ready to accept the connection. We don't want to block our connection since the server pod could crash which we are connecting or in the load-balancing case reject the connection in which case the agent would need to block until timeout. Instead we try sending, error out and try reconnecting.

File: npm/cmd/root.go (hunk 6 above)
Comment (huntergregory) [reply npm/cmd/root.go:22]: perhaps?
https://github.com/Azure/azure-container-networking/blob/d6ba4b9e363cbd394cb544753f7bb0b31fdcf9dc/npm/Dockerfile#L33

Comment (nitishm) [reply npm/cmd/root.go:22]: So are you suggesting we update this to "controller"/"daemon"? The default behavior, as for now, should not use fan-out IMO.

Comment (huntergregory) [reply npm/cmd/root.go:22]: Took another look at the code, and this makes sense now.

**Review by vakalapa — APPROVED** [review]

File: npm/pkg/transport/tls.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rbtr) [reply npm/pkg/transport/tls.go:16]: you could make it dynamic using a volume mount at runtime; it would be better for this to be configurable imo
